metrics-influxdb/src/main/java/io/dropwizard/metrics5/influxdb/InfluxDbUdpSender.java

```diff
@@ -0,0 +1,148 @@
+    private int mtu = 1500;
+
+    private DatagramChannel datagramChannel = null;
+
+    private ByteBuffer byteBuf;
+    private CharBuffer charBuf;
+    private final CharsetEncoder encoder = Charset.forName("UTF-8")
+            .newEncoder()
+            .onMalformedInput(CodingErrorAction.REPLACE)
+            .onUnmappableCharacter(CodingErrorAction.REPLACE);
+
+    /**
+     * Creates a new client which sends data to given address using UDP
+     *
+     * @param hostname The hostname of the InfluxDb server
+     * @param port     The port of the InfluxDb server
+     */
+    public InfluxDbUdpSender(String hostname, int port) {
+        this(new InetSocketAddress(hostname, port));
+    }
+
+    /**
+     * Creates a new client which sends data to given address using UDP
+     *
+     * @param address the address of the InfluxDb server
+     */
+    public InfluxDbUdpSender(InetSocketAddress address) {
+        this.address = Objects.requireNonNull(address);
+        charBuf = CharBuffer.allocate(mtu*2);
+        byteBuf = ByteBuffer.allocate(mtu*2);
+    }
+
+    // for testing
+    void setMTU(int mtu) {
+        this.mtu = mtu;
+    }
+
+    // for testing
+    void setDatagramChannel(DatagramChannel datagramChannel) {
+        this.datagramChannel = datagramChannel;
+    }
+    
+    @Override
+    public void connect() throws IllegalStateException, IOException {
+        if (datagramChannel == null) {
+            datagramChannel = DatagramChannel.open();
+        }
+        byteBuf.clear();
+    }
+
+    @Override
+    public boolean isConnected() {
+        return datagramChannel != null;
+    }
+    
+    @Override
+    public void disconnect() throws IOException {
+        // ignore, keep the datagram channel open
+    }
+
+    @Override
+    public void close() throws IOException {
+        try {
+            datagramChannel.close();
+        } finally {
+            datagramChannel = null;
+        }
+    }
+
+    @Override
+    public void send(StringBuilder str) throws IOException {
+        int len = byteBuf.position();
+        encode(str);
+        int len2 = byteBuf.position();
+        if (len2 >= mtu) {
+            if (len == 0) {
+                // send current buffer (one single measurement exceeds the MTU)
+                sendBuffer();
+                byteBuf.clear();
+            } else {
+                // send previous buffer
+                byteBuf.position(len);
+                sendBuffer();
+                byteBuf.limit(len2);
+                byteBuf.compact();
+            }
+        }
+    }
+    
+    @Override
+    public void flush() throws IOException {
+        if (byteBuf.position() > 0) {
+            sendBuffer();
+            byteBuf.clear();
+        }
+    }
+    
+    private void sendBuffer() throws IOException {
+        byteBuf.flip();
+        datagramChannel.send(byteBuf, address);
+    }
+        
+    private void encode(StringBuilder str) {
```

Comment:
I wonder if all these low-level manipulation are worth. Looks like a lot of code instead of:
```
datagramChannel.send(encoder.encode(CharBuffer.wrap(str)), address)
```

I know this creates a new byte array every time, but the question is whether it's such a big issue. Or maybe we could build a line as a `ByteBuffer` right away instead of `StringBuilder`.

Reply:
In latency sensitive applications, low memory turnover is important. Including a 3rd party library that creates lots of garbage is often not acceptable, since it affects the GC pauses of the entire application.

It would be nice to be able to use (parts of) the metrics library in such an applications. Especially, since collecting metrics is (also) important in latency and performance critical applications.

See for example, the [garbage free implementation of log4j2](https://logging.apache.org/log4j/2.x/manual/garbagefree.html). Their [`TextEncoderHelper`](https://github.com/apache/logging-log4j2/blob/master/log4j-core/src/main/java/org/apache/logging/log4j/core/layout/TextEncoderHelper.java) has more code... as a comparison. I tried to keep this method simple.